clouddriver-titus/src/main/groovy/com/netflix/spinnaker/clouddriver/titus/caching/agents/TitusV2ClusterCachingAgent.java

```diff
@@ -0,0 +1,628 @@
+    List<String> evictFromOnDemandIds = evictFromOnDemand.stream().map(CacheData::getId).collect(Collectors.toList());
+
+    CacheResult result = buildCacheResult(
+      jobs,
+      scalingPolicyResults,
+      allLoadBalancers,
+      taskAndJobIds,
+      onDemandMap,
+      evictFromOnDemandIds
+    );
+
+    result.getCacheResults().get(ON_DEMAND.ns).forEach( onDemandEntry -> {
+      onDemandEntry.getAttributes().put("processedTime", System.currentTimeMillis());
+      onDemandEntry.getAttributes().put("processedCount", Long.parseLong(onDemandEntry.getAttributes().getOrDefault("processedCount", "0").toString()) + 1);
+    });
+
+    PercentileTimer
+      .get(registry, metricId.withTag("operation", "loadData"))
+      .record(System.currentTimeMillis() - startTime, MILLISECONDS);
+    return result;
+  }
+
+  /**
+   * Used to build a cache result, whether normal or on demand,
+   * by transforming known Titus objects into Spinnaker cached objects
+   */
+  private CacheResult buildCacheResult(List<Job> jobs,
+                                       List<ScalingPolicyResult> scalingPolicyResults,
+                                       Map<String, List<String>> allLoadBalancers,
+                                       Map<String, List<String>> taskAndJobIds,
+                                       Map<String, CacheData> onDemandKeep,
+                                       List<String> onDemandEvict) {
+    if (onDemandKeep == null) {
+      onDemandKeep = new HashMap<>();
+    }
+    if (onDemandEvict == null) {
+      onDemandEvict = new ArrayList<>();
+    }
+
+    // INITIALIZE CACHES
+    Map<String, CacheData> applicationCache = createCache();
+    Map<String, CacheData> clusterCache = createCache();
+    Map<String, CacheData> serverGroupCache = createCache();
+    Map<String, CacheData> targetGroupCache = createCache();
+    Map<String, CacheData> imageCache = createCache();
+
+    // Ignore policies in a Deleted state (may need to revisit)
+    List cacheablePolicyStates = Arrays.asList(
+      ScalingPolicyStatus.ScalingPolicyState.Pending,
+      ScalingPolicyStatus.ScalingPolicyState.Applied,
+      ScalingPolicyStatus.ScalingPolicyState.Deleting
+    );
+
+    List<ServerGroupData> serverGroupDatas = jobs.stream()
+      .map( job -> {
+        List<ScalingPolicyData> jobScalingPolicies = scalingPolicyResults.stream()
+          .filter( it -> it.getJobId().equalsIgnoreCase(job.getId()) && cacheablePolicyStates.contains(it.getPolicyState()))
+          .map( it -> new ScalingPolicyData(it.getId().getId(), it.getScalingPolicy(), it.getPolicyState()))
+          .collect(Collectors.toList());
+
+        List<String> jobLoadBalancers = allLoadBalancers.getOrDefault(job.getId(), Collections.emptyList());
+        return new ServerGroupData(job, jobScalingPolicies, jobLoadBalancers, taskAndJobIds.get(job.getId()), account.getName(), region.getName());
+      })
+      .collect(Collectors.toList());
```

Comment:
think you could drop the collect and just chain the following forEach off this stream